### ISSUE TITLE
Deadlock caused by error in BEGIN IMMEDIATE

### ISSUE THREAD

**Issue by kongo2002** [tekartik/sqflite#610]
Hi there,

I am currently seeing some deadlocks by using transactions when the initial call of `BEGIN IMMEDIATE` fails with an exception (here it was `code 5 SQLITE_BUSY`). After that all subsequent queries are blocked due to the failing attempts to get the database lock (`Warning database has been locked for 0:00:10.000000. ...`), I think.

I wasn't able to construct a reproducible example yet because it is not that easy to make `BEGIN` fail predictably.

However, this is the sequence of events that lead up to the 'deadlock':

1. Open of transaction fails unexpectedly
```
03-11 21:19:51.577 12344 12413 I flutter : DatabaseException(database is locked (code 5 SQLITE_BUSY)) sql 'BEGIN IMMEDIATE' args []}
03-11 21:19:51.578 12344 12413 I flutter : #0      wrapDatabaseException (package:sqflite/src/exception_impl.dart:11:7)
03-11 21:19:51.578 12344 12413 I flutter : <asynchronous suspension>
03-11 21:19:51.578 12344 12413 I flutter : #1      SqfliteDatabaseFactoryImpl.wrapDatabaseException (package:sqflite/src/factory_impl.dart:78:7)
03-11 21:19:51.578 12344 12413 I flutter : #2      SqfliteDatabaseMixin.safeInvokeMethod (package:sqflite_common/src/database_mixin.dart:211:15)
03-11 21:19:51.578 12344 12413 I flutter : #3      SqfliteDatabaseMixin.invokeExecute (package:sqflite_common/src/database_mixin.dart:379:12)
03-11 21:19:51.578 12344 12413 I flutter : #4      SqfliteDatabaseMixin.txnExecute.<anonymous closure> (package:sqflite_common/src/database_mixin.dart:371:14)
03-11 21:19:51.578 12344 12413 I flutter : #5      SqfliteDatabaseMixin.txnSynchronized (package:sqflite_common/src/database_mixin.dart:321:26)
03-11 21:19:51.578 12344 12413 I flutter : <asynchronous suspension>
03-11 21:19:51.578 12344 12413 I flutter : #6      SqfliteDatabaseMixin.txnWriteSynchronized (package:sqflite_common/src/database_mixin.dart:354:7)
03-11 21:19:51.578 12344 12413 I flutter : #7      SqfliteDatabaseMixin.txnExecute (package:sqflite_common/src/database_mixin.dart:360:12)
03-11 21:19:51.578 12344 12413 I flutter : #8      SqfliteDatabaseMixin.beginTransaction (package:sqflite_common/src/database_mixin.dart:457:15)
```

2. Transaction calls fail with a `null` reference of the transaction object

```
03-11 21:19:51.610 12344 12413 I flutter : NoSuchMethodError: The method 'rawQuery' was called on null.
03-11 21:19:51.610 12344 12413 I flutter : Receiver: null
03-11 21:19:51.610 12344 12413 I flutter : Tried calling: rawQuery("SELECT a FROM b WHERE c = ? AND d = ? AND e = ?", Instance(length:3) of '_GrowableList')
03-11 21:19:51.611 12344 12413 I flutter : #0      Object.noSuchMethod (dart:core-patch/object_patch.dart:51:5)
[...]
03-11 21:19:51.612 12344 12413 I flutter : NoSuchMethodError: The method 'rawQuery' was called on null.
03-11 21:19:51.612 12344 12413 I flutter : Receiver: null
03-11 21:19:51.612 12344 12413 I flutter : Tried calling: rawQuery("SELECT a FROM z WHERE c = ? AND d = ? AND e = ?", Instance(length:3) of '_GrowableList')
03-11 21:19:51.612 12344 12413 I flutter : #0      Object.noSuchMethod (dart:core-patch/object_patch.dart:51:5)
```

3. All subsequent queries seem to be locked indefinitely

```
03-11 21:20:28.347 12344 12413 I flutter : Warning database has been locked for 0:00:10.000000. Make sure you always use the transaction object for database operations during a transaction
03-11 21:20:35.256 12344 12413 I flutter : Warning database has been locked for 0:00:10.000000. Make sure you always use the transaction object for database operations during a transaction
03-11 21:21:21.047 12344 12413 I flutter : Warning database has been locked for 0:00:10.000000. Make sure you always use the transaction object for database operations during a transaction
[...]
```

I am currently using `sqflite: 1.3.2+3`.

I only had a quick glance at the code yet, so this is just an uneducated guess but I could imagine the error being around `transactionRefCount` (in `_runTransaction`) not being decremented ever in that case. I will try to come up with a test case that could maybe force an error during `beginTransaction` or smth similar but I wasn't able to get the tests running at all so far.

Cheers,
Gregor

**Comment by alextekartik** [comment]
Thanks for the report and analysis. There is indeed a ref count issue. This has also been reported in issue #611 

Fixed in sqflite_common 1.0.3+2 / 2.0.0+2

Can you try a pub pugrade?

**Comment by kongo2002** [comment]
I did some more testing on my end and the deadlock is fixed indeed - thanks for the fast response and the fix!

Cheers,
Gregor
